<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!-- NewPage -->
<html lang="en">
<head>
<!-- Generated by javadoc (1.8.0_202) on Mon Jun 26 11:23:23 EDT 2023 -->
<title>GenomeSizeReference</title>
<meta name="date" content="2023-06-26">
<link rel="stylesheet" type="text/css" href="../../stylesheet.css" title="Style">
<script type="text/javascript" src="../../script.js"></script>
</head>
<body>
<script type="text/javascript"><!--
    try {
        if (location.href.indexOf('is-external=true') == -1) {
            parent.document.title="GenomeSizeReference";
        }
    }
    catch(err) {
    }
//-->
var methods = {"i0":10,"i1":10,"i2":10,"i3":10,"i4":10,"i5":9,"i6":9,"i7":9,"i8":9,"i9":9,"i10":9,"i11":9,"i12":10,"i13":10};
var tabs = {65535:["t0","All Methods"],1:["t1","Static Methods"],2:["t2","Instance Methods"],8:["t4","Concrete Methods"]};
var altColor = "altColor";
var rowColor = "rowColor";
var tableTab = "tableTab";
var activeTableTab = "activeTableTab";
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="topNav"><a name="navbar.top">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.top" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.top.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="class-use/GenomeSizeReference.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../index-files/index-1.html">Index</a></li>
<li><a href="../../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="../../scriptmanager/util/FileSelection.html" title="class in scriptmanager.util"><span class="typeNameLink">Prev&nbsp;Class</span></a></li>
<li><a href="../../scriptmanager/util/GZipUtilities.html" title="class in scriptmanager.util"><span class="typeNameLink">Next&nbsp;Class</span></a></li>
</ul>
<ul class="navList">
<li><a href="../../index.html?scriptmanager/util/GenomeSizeReference.html" target="_top">Frames</a></li>
<li><a href="GenomeSizeReference.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_top">
<li><a href="../../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_top");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
<a name="skip.navbar.top">
<!--   -->
</a></div>
<!-- ========= END OF TOP NAVBAR ========= -->
<!-- ======== START OF CLASS DATA ======== -->
<div class="header">
<div class="subTitle">scriptmanager.util</div>
<h2 title="Class GenomeSizeReference" class="title">Class GenomeSizeReference</h2>
</div>
<div class="contentContainer">
<ul class="inheritance">
<li>java.lang.Object</li>
<li>
<ul class="inheritance">
<li>scriptmanager.util.GenomeSizeReference</li>
</ul>
</li>
</ul>
<div class="description">
<ul class="blockList">
<li class="blockList">
<hr>
<br>
<pre>public class <span class="typeNameLabel">GenomeSizeReference</span>
extends java.lang.Object</pre>
<div class="block">Create a genomic chromosome/contig size reference that is essentially defined
 as an ArrayList chromsome/contig names and a parallel ArrayList storing each
 chromsome/contig's length (in bp). Total bp genome length and genome name are
 also tracked and this class includes helper methods for generating random
 coordinates.</div>
<dl>
<dt><span class="simpleTagLabel">Author:</span></dt>
<dd>William KM Lai</dd>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../scriptmanager/scripts/Peak_Analysis/RandomCoordinate.html" title="class in scriptmanager.scripts.Peak_Analysis"><code>RandomCoordinate</code></a>, 
<a href="../../scriptmanager/scripts/Peak_Analysis/TileGenome.html" title="class in scriptmanager.scripts.Peak_Analysis"><code>TileGenome</code></a></dd>
</dl>
</li>
</ul>
</div>
<div class="summary">
<ul class="blockList">
<li class="blockList">
<!-- ======== CONSTRUCTOR SUMMARY ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor.summary">
<!--   -->
</a>
<h3>Constructor Summary</h3>
<table class="memberSummary" border="0" cellpadding="3" cellspacing="0" summary="Constructor Summary table, listing constructors, and an explanation">
<caption><span>Constructors</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colOne" scope="col">Constructor and Description</th>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../scriptmanager/util/GenomeSizeReference.html#GenomeSizeReference--">GenomeSizeReference</a></span>()</code>
<div class="block">Constructor for empty GenomeSizeReference object.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../scriptmanager/util/GenomeSizeReference.html#GenomeSizeReference-java.lang.String-">GenomeSizeReference</a></span>(java.lang.String&nbsp;build)</code>
<div class="block">Constructor for defining a new instance based on the provided genome build.</div>
</td>
</tr>
</table>
</li>
</ul>
<!-- ========== METHOD SUMMARY =========== -->
<ul class="blockList">
<li class="blockList"><a name="method.summary">
<!--   -->
</a>
<h3>Method Summary</h3>
<table class="memberSummary" border="0" cellpadding="3" cellspacing="0" summary="Method Summary table, listing methods, and an explanation">
<caption><span id="t0" class="activeTableTab"><span>All Methods</span><span class="tabEnd">&nbsp;</span></span><span id="t1" class="tableTab"><span><a href="javascript:show(1);">Static Methods</a></span><span class="tabEnd">&nbsp;</span></span><span id="t2" class="tableTab"><span><a href="javascript:show(2);">Instance Methods</a></span><span class="tabEnd">&nbsp;</span></span><span id="t4" class="tableTab"><span><a href="javascript:show(8);">Concrete Methods</a></span><span class="tabEnd">&nbsp;</span></span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tr id="i0" class="altColor">
<td class="colFirst"><code><a href="../../scriptmanager/objects/CoordinateObjects/GenericCoord.html" title="class in scriptmanager.objects.CoordinateObjects">GenericCoord</a></code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../scriptmanager/util/GenomeSizeReference.html#generateRandomCoord-int-">generateRandomCoord</a></span>(int&nbsp;WINDOW)</code>
<div class="block">Generates a random coordinate of a specified interval length and random
 direction uniformly sampling from genomeSize</div>
</td>
</tr>
<tr id="i1" class="rowColor">
<td class="colFirst"><code>java.util.ArrayList&lt;java.lang.String&gt;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../scriptmanager/util/GenomeSizeReference.html#getChrom--">getChrom</a></span>()</code>
<div class="block">Retrieve list of chromosome names (chromName, parallel to chromSize).</div>
</td>
</tr>
<tr id="i2" class="altColor">
<td class="colFirst"><code>java.util.ArrayList&lt;java.lang.Long&gt;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../scriptmanager/util/GenomeSizeReference.html#getChromSize--">getChromSize</a></span>()</code>
<div class="block">Retrieve list of chromosome lengths in bp (chromSize, parallel to chromName).</div>
</td>
</tr>
<tr id="i3" class="rowColor">
<td class="colFirst"><code>java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../scriptmanager/util/GenomeSizeReference.html#getGenome--">getGenome</a></span>()</code>
<div class="block">Retrieve the name of the current genome build used in this
 GenomeSizeReference instance</div>
</td>
</tr>
<tr id="i4" class="altColor">
<td class="colFirst"><code>long</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../scriptmanager/util/GenomeSizeReference.html#getGenomeSize--">getGenomeSize</a></span>()</code>
<div class="block">Retrieve the total genome size of this GenomeSizeReference instance.</div>
</td>
</tr>
<tr id="i5" class="rowColor">
<td class="colFirst"><code>static void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../scriptmanager/util/GenomeSizeReference.html#initialize_hg19_contig--">initialize_hg19_contig</a></span>()</code>
<div class="block">Set instance to use hg19 (human) genome build.</div>
</td>
</tr>
<tr id="i6" class="altColor">
<td class="colFirst"><code>static void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../scriptmanager/util/GenomeSizeReference.html#initialize_hg19--">initialize_hg19</a></span>()</code>
<div class="block">Set instance to use hg19 (human) genome build.</div>
</td>
</tr>
<tr id="i7" class="rowColor">
<td class="colFirst"><code>static void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../scriptmanager/util/GenomeSizeReference.html#initialize_hg38_contig--">initialize_hg38_contig</a></span>()</code>
<div class="block">Set instance to use hg38/GrCh38 (human) genome build.</div>
</td>
</tr>
<tr id="i8" class="altColor">
<td class="colFirst"><code>static void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../scriptmanager/util/GenomeSizeReference.html#initialize_hg38--">initialize_hg38</a></span>()</code>
<div class="block">Set instance to use hg38/GrCh38 (human) genome build.</div>
</td>
</tr>
<tr id="i9" class="rowColor">
<td class="colFirst"><code>static void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../scriptmanager/util/GenomeSizeReference.html#initialize_mm10--">initialize_mm10</a></span>()</code>
<div class="block">Set instance to use mm10 (mouse) genome build.</div>
</td>
</tr>
<tr id="i10" class="altColor">
<td class="colFirst"><code>static void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../scriptmanager/util/GenomeSizeReference.html#initialize_sacCer3_cegr--">initialize_sacCer3_cegr</a></span>()</code>
<div class="block">Set instance to use sacCer3 (yeast) genome build using the *arabic* numeral
 chromosome naming system.</div>
</td>
</tr>
<tr id="i11" class="rowColor">
<td class="colFirst"><code>static void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../scriptmanager/util/GenomeSizeReference.html#initialize_sacCer3--">initialize_sacCer3</a></span>()</code>
<div class="block">Set instance to use sacCer3 (yeast) genome build using the *roman* numeral
 chromosome naming system.</div>
</td>
</tr>
<tr id="i12" class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../scriptmanager/util/GenomeSizeReference.html#isSmaller-int-">isSmaller</a></span>(int&nbsp;WINDOW)</code>
<div class="block">Check to make sure WINDOW size is smaller than every chromosome to prevent
 infinite loop.</div>
</td>
</tr>
<tr id="i13" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../scriptmanager/util/GenomeSizeReference.html#setGenome-java.lang.String-">setGenome</a></span>(java.lang.String&nbsp;build)</code>
<div class="block">Initialize genome build to a predefined set of chromosome/contig sets.</div>
</td>
</tr>
</table>
<ul class="blockList">
<li class="blockList"><a name="methods.inherited.from.class.java.lang.Object">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;java.lang.Object</h3>
<code>equals, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</code></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="details">
<ul class="blockList">
<li class="blockList">
<!-- ========= CONSTRUCTOR DETAIL ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor.detail">
<!--   -->
</a>
<h3>Constructor Detail</h3>
<a name="GenomeSizeReference--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>GenomeSizeReference</h4>
<pre>public&nbsp;GenomeSizeReference()</pre>
<div class="block">Constructor for empty GenomeSizeReference object. I.e., empty chromName &amp;
 chromSize lists with genomeSize set to 0 and genomeName as an empty Sstring.</div>
</li>
</ul>
<a name="GenomeSizeReference-java.lang.String-">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>GenomeSizeReference</h4>
<pre>public&nbsp;GenomeSizeReference(java.lang.String&nbsp;build)</pre>
<div class="block">Constructor for defining a new instance based on the provided genome build.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>build</code> - indicate the genome build to use</dd>
</dl>
</li>
</ul>
</li>
</ul>
<!-- ============ METHOD DETAIL ========== -->
<ul class="blockList">
<li class="blockList"><a name="method.detail">
<!--   -->
</a>
<h3>Method Detail</h3>
<a name="setGenome-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setGenome</h4>
<pre>public&nbsp;void&nbsp;setGenome(java.lang.String&nbsp;build)
               throws java.lang.IllegalArgumentException</pre>
<div class="block">Initialize genome build to a predefined set of chromosome/contig sets.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>build</code> - this string determines which genome build to use (see initialize_...() methods)</dd>
<dt><span class="throwsLabel">Throws:</span></dt>
<dd><code>java.lang.IllegalArgumentException</code> - when unrecognized build String is input, this exception is thrown</dd>
</dl>
</li>
</ul>
<a name="getGenome--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getGenome</h4>
<pre>public&nbsp;java.lang.String&nbsp;getGenome()</pre>
<div class="block">Retrieve the name of the current genome build used in this
 GenomeSizeReference instance</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>a String describing the current genome build (e.g. sacCer3, hg19_contigs, ...)</dd>
</dl>
</li>
</ul>
<a name="getGenomeSize--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getGenomeSize</h4>
<pre>public&nbsp;long&nbsp;getGenomeSize()</pre>
<div class="block">Retrieve the total genome size of this GenomeSizeReference instance.</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>genomeSize which should be equivalent to sum of all Longs in chromSize</dd>
</dl>
</li>
</ul>
<a name="getChrom--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getChrom</h4>
<pre>public&nbsp;java.util.ArrayList&lt;java.lang.String&gt;&nbsp;getChrom()</pre>
<div class="block">Retrieve list of chromosome names (chromName, parallel to chromSize).</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>list of chromosome names (Strings)</dd>
</dl>
</li>
</ul>
<a name="getChromSize--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getChromSize</h4>
<pre>public&nbsp;java.util.ArrayList&lt;java.lang.Long&gt;&nbsp;getChromSize()</pre>
<div class="block">Retrieve list of chromosome lengths in bp (chromSize, parallel to chromName).</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>list of bp lengths (Long)</dd>
</dl>
</li>
</ul>
<a name="initialize_sacCer3--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>initialize_sacCer3</h4>
<pre>public static&nbsp;void&nbsp;initialize_sacCer3()</pre>
<div class="block">Set instance to use sacCer3 (yeast) genome build using the *roman* numeral
 chromosome naming system. This includes the "complete" assembled chromosomes,
 the mitochondrial chromosome ("chrM"), and the 2-micron plasmid.

 https://hgdownload.soe.ucsc.edu/goldenPath/sacCer3/bigZips/sacCer3.chrom.sizes</div>
</li>
</ul>
<a name="initialize_sacCer3_cegr--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>initialize_sacCer3_cegr</h4>
<pre>public static&nbsp;void&nbsp;initialize_sacCer3_cegr()</pre>
<div class="block">Set instance to use sacCer3 (yeast) genome build using the *arabic* numeral
 chromosome naming system. This includes the "complete" assembled chromosomes,
 the mitochondrial chromosome ("chrM"), and the 2-micron plasmid.

 https://hgdownload.soe.ucsc.edu/goldenPath/sacCer3/bigZips/sacCer3.chrom.sizes</div>
</li>
</ul>
<a name="initialize_hg19--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>initialize_hg19</h4>
<pre>public static&nbsp;void&nbsp;initialize_hg19()</pre>
<div class="block">Set instance to use hg19 (human) genome build. This includes only the
 "complete" assembled chromosomes. Excludes the mitochondrial chromosome and
 alternative contigs.

 https://hgdownload.soe.ucsc.edu/goldenPath/hg19/bigZips/hg19.chrom.sizes</div>
</li>
</ul>
<a name="initialize_hg19_contig--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>initialize_hg19_contig</h4>
<pre>public static&nbsp;void&nbsp;initialize_hg19_contig()</pre>
<div class="block">Set instance to use hg19 (human) genome build. This includes the "complete"
 assembled chromosomes, the mitochondrial chromosome ("chrM"), and the full
 set of alternative contigs.

 https://hgdownload.soe.ucsc.edu/goldenPath/hg19/bigZips/hg19.chrom.sizes</div>
</li>
</ul>
<a name="initialize_hg38--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>initialize_hg38</h4>
<pre>public static&nbsp;void&nbsp;initialize_hg38()</pre>
<div class="block">Set instance to use hg38/GrCh38 (human) genome build. This includes only the
 "complete" assembled chromosomes. Excludes the mitochondrial chromosome and
 alternative contigs.

 https://github.com/igvteam/igv/blob/master/genomes/sizes/hg38.chrom.sizes</div>
</li>
</ul>
<a name="initialize_hg38_contig--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>initialize_hg38_contig</h4>
<pre>public static&nbsp;void&nbsp;initialize_hg38_contig()</pre>
<div class="block">Set instance to use hg38/GrCh38 (human) genome build. This includes the
 "complete" assembled chromosomes, the mitochondrial chromosome ("chrM"), and
 the full set of alternative contigs.

 https://github.com/igvteam/igv/blob/master/genomes/sizes/hg38.chrom.sizes</div>
</li>
</ul>
<a name="initialize_mm10--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>initialize_mm10</h4>
<pre>public static&nbsp;void&nbsp;initialize_mm10()</pre>
<div class="block">Set instance to use mm10 (mouse) genome build. This includes only the
 "complete" assembled chromosomes. Excludes the mitochondrial chromosome and
 alternative contigs.

 https://hgdownload.soe.ucsc.edu/goldenPath/mm10/bigZips/mm10.chrom.sizes</div>
</li>
</ul>
<a name="generateRandomCoord-int-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>generateRandomCoord</h4>
<pre>public&nbsp;<a href="../../scriptmanager/objects/CoordinateObjects/GenericCoord.html" title="class in scriptmanager.objects.CoordinateObjects">GenericCoord</a>&nbsp;generateRandomCoord(int&nbsp;WINDOW)</pre>
<div class="block">Generates a random coordinate of a specified interval length and random
 direction uniformly sampling from genomeSize</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>WINDOW</code> - window size (interval length) of returned coordinate interval</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>random coordinate interval with chr, start, stop, and direction</dd>
</dl>
</li>
</ul>
<a name="isSmaller-int-">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>isSmaller</h4>
<pre>public&nbsp;boolean&nbsp;isSmaller(int&nbsp;WINDOW)</pre>
<div class="block">Check to make sure WINDOW size is smaller than every chromosome to prevent
 infinite loop.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>WINDOW</code> - the size of the window to check</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>true if WINDOW is smaller than all chr, false if there is at least one chr smaller than WINDOW</dd>
</dl>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
<!-- ========= END OF CLASS DATA ========= -->
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottomNav"><a name="navbar.bottom">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.bottom" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.bottom.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="class-use/GenomeSizeReference.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../index-files/index-1.html">Index</a></li>
<li><a href="../../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="../../scriptmanager/util/FileSelection.html" title="class in scriptmanager.util"><span class="typeNameLink">Prev&nbsp;Class</span></a></li>
<li><a href="../../scriptmanager/util/GZipUtilities.html" title="class in scriptmanager.util"><span class="typeNameLink">Next&nbsp;Class</span></a></li>
</ul>
<ul class="navList">
<li><a href="../../index.html?scriptmanager/util/GenomeSizeReference.html" target="_top">Frames</a></li>
<li><a href="GenomeSizeReference.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_bottom">
<li><a href="../../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_bottom");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li>Field&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
<a name="skip.navbar.bottom">
<!--   -->
</a></div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
</body>
</html>
